### PR TITLE
[Rule] Over Taunt Hate

### PR DESCRIPTION
**Pull request by fryguy503**
Rule to add additional hate when taunt succeeds. This can help tune snap aggro on taunt classes. Only works when taunt succeeds and not already top hate.

### CODE REVIEW

File: zone/special_attacks.cpp [hunk 1]
```diff
@@ -2129,68 +2129,70 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
 void Mob::Taunt(NPC *who, bool always_succeed, int chance_bonus, bool FromSpell, int32 bonus_hate)
 {
-	if (who == nullptr)
-		return;
-
-	if (DivineAura())
-		return;
-
-	if (!FromSpell && !CombatRange(who))
+	if (who == nullptr || DivineAura() || (!FromSpell && !CombatRange(who))) {
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2132]: Can do `!who` and change `FromSpell` to `from_spell`.

File: zone/special_attacks.cpp [hunk 2]
```diff
@@ -2129,68 +2129,70 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
 	Mob *hate_top = who->GetHateMost();
 
 	int level_difference = GetLevel() - who->GetLevel();
 	bool success = false;
 
 	// Support for how taunt worked pre 2000 on LIVE - Can not taunt NPC over your level.
-	if ((RuleB(Combat, TauntOverLevel) == false) && (level_difference < 0) ||
-	    who->GetSpecialAbility(IMMUNE_TAUNT)) {
+	if (
+		RuleB(Combat, TauntOverLevel) == false &&
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2147]: `!RuleB(Combat, TauntOverLevel)`

File: zone/special_attacks.cpp [hunk 3]
```diff
@@ -2129,68 +2129,70 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
+	) {
 		MessageString(Chat::SpellFailure, FAILED_TAUNT);
 		return;
 	}
 
 	// All values used based on live parses after taunt was updated in 2006.
-	if ((hate_top && hate_top->GetHPRatio() >= 20) || hate_top == nullptr || chance_bonus) {
+	if (
+		(hate_top && hate_top->GetHPRatio() >= 20) ||
+		hate_top == nullptr ||
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2158]: `!hate_top`

File: zone/special_attacks.cpp [hunk 4]
```diff
@@ -2129,68 +2129,70 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
+			if (IsClient() && !always_succeed) {
 				tauntchance -= (RuleR(Combat, TauntSkillFalloff) *
 						(CastToClient()->MaxSkill(EQ::skills::SkillTaunt) -
 						 GetSkill(EQ::skills::SkillTaunt)));
+			}
 
-			if (tauntchance < 1)
+			if (tauntchance < 1) {
 				tauntchance = 1.0f;
+			}
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2194]: `taunt_chance`

File: zone/special_attacks.cpp [hunk 5]
```diff
@@ -2205,17 +2207,18 @@ void Mob::Taunt(NPC *who, bool always_succeed, int chance_bonus, bool FromSpell,
 
 		if (success) {
 			if (hate_top && hate_top != this) {
-				int64 newhate = (who->GetNPCHate(hate_top) - who->GetNPCHate(this)) + 1 + bonus_hate;
+				int64 newhate = ((who->GetNPCHate(hate_top) - who->GetNPCHate(this)) + bonus_hate + RuleI(Combat, TauntOverAggro));
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2210]: `new_hate`

**Review by Kinglykrab — APPROVED** [review]

File: zone/special_attacks.cpp [hunk 6]
```diff
@@ -2127,105 +2127,109 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
+				taunt_chance,
 				success ? "true" : "false",
 				hate_top ? hate_top->GetMobDescription() : "not found"
 			);
 		}
 
 		if (success) {
 			if (hate_top && hate_top != this) {
-				int64 newhate = (who->GetNPCHate(hate_top) - who->GetNPCHate(this)) + 1 + bonus_hate;
+				int64 new_hate = ((who->GetNPCHate(hate_top) - who->GetNPCHate(this)) + bonus_hate + RuleI(Combat, TauntOverAggro));
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2210]: We lost the `+ 1` here, I think that's just so we always have at least `1` hate.

Comment (fryguy503) [reply zone/special_attacks.cpp:2210]: Not there, I should have it as the default TauntOverAggro rule no need to have a lingering +1

File: zone/special_attacks.cpp [hunk 7]
```diff
@@ -2127,105 +2127,114 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
 	int level_difference = GetLevel() - who->GetLevel();
 	bool success = false;
 
 	// Support for how taunt worked pre 2000 on LIVE - Can not taunt NPC over your level.
-	if ((RuleB(Combat, TauntOverLevel) == false) && (level_difference < 0) ||
-	    who->GetSpecialAbility(IMMUNE_TAUNT)) {
+	if (
+		!RuleB(Combat, TauntOverLevel) && 
+		level_difference < 0 ||
+	    who->GetSpecialAbility(IMMUNE_TAUNT)
```
Comment (Akkadius) [inline zone/special_attacks.cpp:2149]: Indent not consistent with other lines

**Pull request merged**